File: product-of-array-except-self/product-of-array-except-self.c


/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* productExceptSelf(int* nums, int numsSize, int* returnSize){
    
    if(numsSize==0) return 0;
    
    *returnSize=numsSize;
    
    int *result=(int*)malloc(sizeof(int)*numsSize);  //Array declaration
    
    int l1=1;
    int r1=1;
    
    int *lm=(int*)malloc(sizeof(int)*numsSize);  //Array declaration
    int *rm=(int*)malloc(sizeof(int)*numsSize);  //Array declaration
    
    int lprod=1;
    int rprod=1;
    
    for(int i=0;i< numsSize;i++){
        lprod=l1*lprod;
        lm[i]=lprod;
        l1=nums[i];
    }
    
    for(int i=numsSize-1;i>=0;i--){
        rprod=r1*rprod;
        rm[i]=rprod;
        r1=nums[i];
    }
    
    for(int i=numsSize-1;i>=0;i--){
        result[i]=lm[i]*rm[i];
    }
    
    return result;

    
    
    
    
    

}

File: product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        
        answer=[0]*len(nums)
        l=[0]*len(nums)
        r=[0]*len(nums)
        
        l[0]=1
        r[len(nums)-1]=1
        lprod=1
        rprod=1
        
        
        for i in range(1,len(nums)):
            lprod=nums[i-1]*lprod
            l[i]=lprod
            print(l[i])
            
        for i in range(len(nums)-2,-1,-1):
            rprod=nums[i+1]*rprod
            r[i]=rprod
            print (r[i])
            
        for i in range(0,len(nums)):
            answer[i]=l[i]*r[i]
            
        return answer